File: src/test/benchmark.cpp
#include <iostream>
#include <vector>
#include <string>
#include <chrono>
#include <thread>
#include <atomic>
#include <mutex>
#include <algorithm>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <iomanip>
#include <netdb.h>

struct BenchResult {
    size_t total_ops;
    std::vector<double> latencies; // in milliseconds
    double total_time;
};

class BenchClient {
public:
    BenchClient(std::string host, int port) : host_(host), port_(port), sock_(-1) {}

    bool connect() {
        struct addrinfo hints{}, *res;
        hints.ai_family = AF_INET;
        hints.ai_socktype = SOCK_STREAM;
        if (getaddrinfo(host_.c_str(), std::to_string(port_).c_str(), &hints, &res) != 0) return false;

        sock_ = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
        if (sock_ < 0) { freeaddrinfo(res); return false; }

        if (::connect(sock_, res->ai_addr, res->ai_addrlen) < 0) {
            close(sock_);
            freeaddrinfo(res);
            return false;
        }
        freeaddrinfo(res);
        return true;
    }

    void close_conn() {
        if (sock_ != -1) close(sock_);
    }

    bool send_cmd(const std::string& cmd) {
        std::string full_cmd = cmd + "\n";
        ssize_t total_sent = 0;
        while (total_sent < (ssize_t)full_cmd.size()) {
            ssize_t sent = send(sock_, full_cmd.c_str() + total_sent, full_cmd.size() - total_sent, 0);
            if (sent <= 0) return false;
            total_sent += sent;
        }

        char buf[4096];
        ssize_t r = recv(sock_, buf, sizeof(buf), 0);
        return r > 0;
    }

private:
    std::string host_;
    int port_;
    int sock_;
};

void run_worker(std::string host, int port, int ops_per_thread, std::string cmd_template, BenchResult& res) {
    BenchClient client(host, port);
    if (!client.connect()) return;

    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < ops_per_thread; ++i) {
        std::string cmd = cmd_template;
        size_t pos = cmd.find("{}");
        if (pos != std::string::npos) {
            cmd.replace(pos, 2, std::to_string(i));
        }

        auto op_start = std::chrono::high_resolution_clock::now();
        if (!client.send_cmd(cmd)) break;
        auto op_end = std::chrono::high_resolution_clock::now();

        std::chrono::duration<double, std::milli> diff = op_end - op_start;
        res.latencies.push_back(diff.count());
    }
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> total = end - start;
    res.total_time = total.count();
    res.total_ops = res.latencies.size();
    client.close_conn();
}

void print_stats(const std::string& title, std::vector<BenchResult>& results) {
    size_t total_ops = 0;
    std::vector<double> all_latencies;
    double max_time = 0;

    for (auto& r : results) {
        total_ops += r.total_ops;
        all_latencies.insert(all_latencies.end(), r.latencies.begin(), r.latencies.end());
        max_time = std::max(max_time, r.total_time);
    }

    if (all_latencies.empty()) return;

    std::sort(all_latencies.begin(), all_latencies.end());
    double p50 = all_latencies[all_latencies.size() * 0.5];
    double p95 = all_latencies[all_latencies.size() * 0.95];
    double p99 = all_latencies[all_latencies.size() * 0.99];
    double avg = 0;
    for (double l : all_latencies) avg += l;
    avg /= all_latencies.size();

    std::cout << "\n=== " << title << " ===\n";
    std::cout << "Throughput: " << std::fixed << std::setprecision(0) << (total_ops / max_time) << " ops/sec\n";
    std::cout << "Latency (ms): \n";
    std::cout << "  Avg: " << std::fixed << std::setprecision(3) << avg << "\n";
    std::cout << "  p50: " << p50 << "\n";
    std::cout << "  p95: " << p95 << "\n";
    std::cout << "  p99: " << p99 << "\n";
}

int main(int argc, char** argv) {
    std::string host = "localhost";
    int port = 5000;
    int threads = 4;
    int ops = 10000;

    if (argc > 1) host = argv[1];
    if (argc > 2) port = std::stoi(argv[2]);
    if (argc > 3) threads = std::stoi(argv[3]);
    if (argc > 4) ops = std::stoi(argv[4]);

    std::cout << "Benchmarking KVick-DB at " << host << ":" << port << " with " << threads << " threads, " << ops << " total ops\n";

    int ops_per_thread = ops / threads;
    std::vector<BenchResult> set_results(threads);
    std::vector<std::thread> workers;

    // SET Benchmark
    for (int i = 0; i < threads; ++i) {
        workers.emplace_back(run_worker, host, port, ops_per_thread, "SET key_{} val_{}", std::ref(set_results[i]));
    }
    for (auto& t : workers) t.join();
    print_stats("SET (W=2)", set_results);

    // GET Benchmark
    workers.clear();
    std::vector<BenchResult> get_results(threads);
    for (int i = 0; i < threads; ++i) {
        workers.emplace_back(run_worker, host, port, ops_per_thread, "GET key_{}", std::ref(get_results[i]));
    }
    for (auto& t : workers) t.join();
    print_stats("GET (R=2)", get_results);

    return 0;
}

File: src/test/benchmark.py
#!/usr/bin/env python3
import socket
import time
import threading
import statistics
import argparse
import re
from concurrent.futures import ThreadPoolExecutor

class KVickBench:
    def __init__(self, host, port, threads, ops):
        self.host = host
        self.port = port
        self.threads = threads
        self.ops = ops
        self.latencies = []
        self.node_counts = {}
        self.lock = threading.Lock()
        self.total_success = 0
        self.sample_mapping = None

    def run_worker(self, ops_per_thread, cmd_template):
        worker_latencies = []
        worker_nodes = {}
        success_count = 0
        
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.connect((self.host, self.port))
            
            for i in range(ops_per_thread):
                cmd = cmd_template.format(i) + "\n"
                
                start = time.perf_counter()
                sock.sendall(cmd.encode())
                
                resp = b""
                while b"\n" not in resp:
                    chunk = sock.recv(1024)
                    if not chunk: break
                    resp += chunk
                
                end = time.perf_counter()
                
                if resp:
                    resp_str = resp.decode()
                    worker_latencies.append((end - start) * 1000) # ms
                    success_count += 1
                    
                    # Extract node IDs from context={"node1":1, ...}
                    found_nodes = re.findall(r'"([^"]+)":', resp_str)
                    for node in found_nodes:
                        worker_nodes[node] = worker_nodes.get(node, 0) + 1
                    
                    # Capture the first successful mapping for display
                    if self.sample_mapping is None and "bench_key_0" in cmd:
                        self.sample_mapping = (cmd.strip(), found_nodes)
            
            sock.close()
        except Exception as e:
            print(f"Worker error: {e}")

        with self.lock:
            self.latencies.extend(worker_latencies)
            self.total_success += success_count
            for node, count in worker_nodes.items():
                self.node_counts[node] = self.node_counts.get(node, 0) + count

    def print_stats(self, title, total_time):
        if not self.latencies:
            print(f"\n=== {title} ===\nNo data collected.")
            return

        self.latencies.sort()
        avg_lat = sum(self.latencies) / len(self.latencies)
        p50 = statistics.median(self.latencies)
        p95 = self.latencies[int(len(self.latencies) * 0.95)]
        p99 = self.latencies[int(len(self.latencies) * 0.99)]
        throughput = self.total_success / total_time

        print(f"\n=== {title} ===")
        if self.sample_mapping:
            key_cmd, nodes = self.sample_mapping
            print(f"Sample Mapping: {key_cmd} -> Quorum: {', '.join(nodes)}")
            self.sample_mapping = None # Clear for next phase

        print(f"Throughput:  {throughput:.2f} ops/sec")
        print(f"Latency (ms):")
        print(f"  Avg:       {avg_lat:.3f}")
        print(f"  P50:       {p50:.3f}")
        print(f"  P95:       {p95:.3f}")
        print(f"  P99:       {p99:.3f}")
        
        print(f"Node Distribution (Quorum Appearances):")
        total_hits = sum(self.node_counts.values())
        for node in sorted(self.node_counts.keys()):
            count = self.node_counts[node]
            pct = (count / total_hits) * 100 if total_hits > 0 else 0
            print(f"  {node:10}: {count:6} hits ({pct:4.1f}%)")
        self.node_counts = {} # Reset for next phase

    def run(self):
        print(f"Benchmarking KVick-DB at {self.host}:{self.port}")
        print(f"Threads: {self.threads}, Ops per phase: {self.ops}")

        ops_per_thread = self.ops // self.threads

        # --- SET Phase ---
        self.latencies = []
        self.total_success = 0
        start_time = time.perf_counter()
        
        with ThreadPoolExecutor(max_workers=self.threads) as executor:
            for _ in range(self.threads):
                executor.submit(self.run_worker, ops_per_thread, "SET bench_key_{0} bench_val_{0}")
        
        end_time = time.perf_counter()
        self.print_stats("SET (W=2)", end_time - start_time)

        # --- GET Phase ---
        self.latencies = []
        self.total_success = 0
        start_time = time.perf_counter()
        
        with ThreadPoolExecutor(max_workers=self.threads) as executor:
            for _ in range(self.threads):
                executor.submit(self.run_worker, ops_per_thread, "GET bench_key_{}")
        
        end_time = time.perf_counter()
        self.print_stats("GET (R=2)", end_time - start_time)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="KVick-DB Python Benchmark")
    parser.add_argument("--host", default="localhost", help="Host address (default: localhost)")
    parser.add_argument("--port", type=int, default=5001, help="Port (default: 5001)")
    parser.add_argument("--threads", type=int, default=8, help="Number of concurrent threads (default: 8)")
    parser.add_argument("--ops", type=int, default=10000, help="Total operations per phase (default: 10000)")
    
    args = parser.parse_args()
    
    bench = KVickBench(args.host, args.port, args.threads, args.ops)
    bench.run()
